tests/production/all-tests.js

```diff
@@ -0,0 +1,34 @@
+/*!
+GPII Universal Tests
+
+Copyright 2019 OCAD University
+
+Licensed under the New BSD license. You may not use this file except in
+compliance with this License.
+
+You may obtain a copy of the License at
+https://github.com/GPII/universal/blob/master/LICENSE.txt
+*/
+
+"use strict";
+
+var fluid = require("infusion"),
+    kettle = fluid.require("kettle");
+
+// Ensure this happens first, to catch errors during code loading.
+kettle.loadTestingSupport();
+
+// Pass the current `require` to `fluid.require`, as nyc's instrumentation is hooked into it.
+fluid.require("%gpii-universal", require);
+
+var testIncludes = [
+    "./AddUserSettingsToCouchTests.js",
```

Comment:
This first step that inserts user keys into the database and the last step at line 29 that removes these keys are only required by a couple of test definitions. I'm thinking it might be worthwhile to refactor these 2 steps into test sequence elements that can be included in test sequences when needed. See [Fixtures.js](https://github.com/GPII/universal/blob/master/gpii/node_modules/testing/src/Fixtures.js) for examples of test sequence elements and how they are used.

@amb26, do you think if it's worthwhile to make this change?

Another concern is, if a test sequence includes these steps that adds user keys at the start and deletes them at the end, what if the test fails in the middle (after user keys are added but before reaching the deleting step)? @amb26, is there an event that a listener can be attached to perform the clean up (deletions in this case) if the test fails unexpectedly?

Reply:
We should just have a generalised cleanup that guarantees to undo all effects of the tests on the database, regardless of their disposition

Reply:
The "gpii" database is already reprovisioned by the caseHolder regardless of whether the previous test suite failed or succeeded.

Reply:
@amb26 As a last step there are tests that involve requests to delete these test keys and prefssafes from the database.  As it happens, the latest merge with master causes some of the other tests in the suite to fail.  Nonetheless, these last step deletion tests run and succeed.  Is that consistent with the kind of generalised cleanup that you are recommending?

Reply:
@the-t-in-rtf, The tests discussed here involve the production database.  If I understand you, I would hope that running theses tests would not re-provision that database.  All that happens is that a few gpii 'user' keys and prefssafes are added to the database, tests are run using them, and then they are deleted.  Other than that the database should not be affected.

Reply:
Thanks for clarifying, the whole concept of "production tests" gets a bit muddier than I'd like, but of course we do have to test our production configurations somehow.  I guess we used to use the pouch-reset endpoint for this, but we should never have that available in a production config.  If you wanted, you could make a "productionTests" config that mixes in the required db provisioning.  That would clearly be different than the production config, and would avoid having to remember and undo every change every time.

Reply:
Thanks @the-t-in-rtf.  The "production tests" have never used pouch, at least since I started working on GPII-3333.  A docker container is run with a provisioned couch data base in the same way as it is in actual production.  To be complete:  three linked containers are running:  one for the cloud-based flowmanager, one for the preferences server and one for couch.  (Actually there's a fourth container that runs the tests themselves).  The docker images that are used to create these containers are the very same images used in actual production.

Reply:
@cindyli 

There are two types of failures when it comes to running these tests.

One kind is an "assertion failure" where some `jqUnitXXX()` test function is executed and results in an exception or test failure.  In this case, the system moves on to the next test and execution continues until the final "jq: All tests concluded:" message is output.  This kind of exception is handled by the testing framework itself such that subsequent tests and code continue to run.

The order that the production configuration tests are run is such that the tests that remove the extra keys and prefs safes are executed as the last step, meaning they will be run even if some previous test(s) fail(s).  Also, the database cleanup code is sensitive to database requests of missing records such that no test failure occurs for any missing record.  Only requests for the extra test keys and prefs safe that return a record lead to their deletion.

Another kind of failure is "catastrophic" failure. In this case, some exception causes the entire test machinery to shut down and any remaining tests are not executed.  This situation does mean the last step database cleanup tests are not executed.  (Pause:  could such code make it into master?)  One way to accommodate this kind of failure is to catch the exception that causes catastrophic failure and then run the deletion tests.

The latest push uses a `try/catch` block for this purpose.  I've experimented with forcing a test failure and a catastrophic failure using `fluid.fail()` and the data base cleanup is executed in both cases.  Let me know if there is a better way.

Push coming up.

Reply:
@amb26 can you comment on [the above](https://github.com/GPII/universal/pull/718#discussion_r267075225), please?  Thanks.

Reply:
The try-catch block doesn't seem to hold much value since almost the entirety of the test sequence operates asynchronously. I suggest the best approach is to add the cleanup in the following two places:
i) add a sequenceGrade element to every test fixture with priority: "last:testing" which executes the cleanup as the last ordinary action of every test sequence
ii) unfortunately neither QUnit nor jQunit allow for asynchronous cleanup actions. What I've found is the following module that will let you install this at the process level - https://github.com/jtlapp/node-cleanup - I suggest you use the pattern in the docs introduced as "You may perform asynchronous cleanup upon receiving a signal, as follows:"

Reply:
It struck me that there'll be a problem running a test sequence as part of this cleanup callback - since it will kick off another round of jqUnit execution with its own invocation of process.exit() as written here - https://github.com/fluid-project/node-jqunit/blob/master/lib/jqUnit-node.js#L175 - which the author of node-cleanup warns us explicitly not to do.

One route would be to rewrite DeleteUserSettingsFromCouchTests.js so that it simply consists of a sequence of promises as we have written in the DataLoader at https://github.com/GPII/universal/blob/master/scripts/deleteAndLoadSnapsets.js#L352 .

Another option would be to write a monkey-patch to jqUnit.allTestsDone within your node-cleanup callback which eliminates all its reporting and is the site responsible for calling process.kill as recommended.

I'm not sure I see a clear front-runner amongst these options, which are indeed both rather peculiar, reflecting the fact that "Production Tests" is itself a rather peculiar concept. I'm happy for you to choose either of them, I think.

Reply:
Thinking about this a bit more, I feel that both of these suggestions are nuts. I recommend instead that you package the deletion tests as a separately runnable command-line task - that is, that you make DeleteUserSettingsFromCouchTests.js self-running, and invoke it as a cleanup step separately from the npm scripts level. In this case there is the further advantage that if there is some other kind of failure, it is clear to the integration team how to trigger cleanup manually.

Reply:
For anyone who was as confused and intrigued as I was by the whole concept of `last:testing`, [here are the relevant docs](https://docs.fluidproject.org/infusion/development/Priorities.html#supported-values-for-priorities).

Reply:
Thanks @the-t-in-rtf.  I began intrigued, and did a recursive grep on all of the modules within universal's gpii.  I found nothing, became confused, and decided to have lunch.  (I was about to do another grep on infusion's source when your email arrived).

Reply:
@amb26 I like your latest suggestion (make `DeleteUserSettingsFromCouchTests.js` self-running), and am leaning toward modifying [`vagrantCloudBasedContainers.sh`](https://github.com/GPII/universal/blob/master/scripts/vagrantCloudBasedContainers.sh).  That script is what sets everything up for the production configuration tests, and then executes them from within a "production config" docker container.  I'll think on it.

In addition, I have a nagging feeling that something similar is needed for adding the extra keys/prefssafes, see `AddUserSettingsForCouchTests.js` -- that it is guaranteed to run first and provision the database before the tests that need them run.   Then again, if that addition fails then the dependent tests will fail, and there is nothing to delete in the end.   Hmmmm.

Reply:
@amb26 wrote:

> ...
> you make DeleteUserSettingsFromCouchTests.js self-running, and invoke it as a cleanup step separately from the npm scripts level.
> ...
Done with https://github.com/GPII/universal/pull/718#commits-pushed-6fe7b0a